packages/cozy-realtime/src/index.js

```diff
@@ -1,339 +1,320 @@
-            )
-          }
-        }, retryDelay)
-      } else {
-        console.error(`0 tries left. Stop reconnecting realtime.`)
-        // remove cached socket and promise
-        if (socketPromise) socketPromise = null
-        if (cozySocket) cozySocket = null
-      }
-    }
+  /**
+   * Update token on socket
+   */
+  _updateAuthentication() {
+    logger.info('Update token on socket')
+    this._socket.authenticate()
   }
 
-  createWebSocket(
-    config,
-    onSocketMessage,
-    onSocketClose,
-    NUM_RETRIES,
-    RETRY_BASE_DELAY
-  )
-
-  cozySocket = {
-    subscribe: (doctype, event, listener, docId) => {
-      if (typeof listener !== 'function')
-        throw new Error('Realtime event listener must be a function')
-
-      const listenerKey = getListenerKey(doctype, docId)
-
-      if (!listeners.has(listenerKey)) {
-        listeners.set(listenerKey, {})
-        subscribeWhenReady(doctype, docId)
-      }
-
-      const eventListeners = listeners.get(listenerKey)[event] || []
-      eventListeners.push(listener)
-      listeners.set(listenerKey, {
-        ...listeners.get(listenerKey),
-        [event]: eventListeners
-      })
-    },
-    unsubscribe: (doctype, event, listener, docId) => {
-      const listenerKey = getListenerKey(doctype, docId)
-      if (listeners.has(listenerKey)) {
-        const socketListeners = listeners.get(listenerKey)
-        if (
-          socketListeners[event] &&
-          socketListeners[event].includes(listener)
-        ) {
-          listeners.set(listenerKey, {
-            ...socketListeners,
-            [event]: socketListeners[event].filter(l => l !== listener)
-          })
-        }
-        if (!hasListeners(listeners.get(listenerKey))) {
-          listeners.delete(listenerKey)
-        }
-      }
+  /**
+   * Reset socket
+   */
+  _resetSocket() {
+    if (this._socket) {
+      this._socket.close()
+      this._socket = null
     }
+    this._createSocket()
   }
-}
 
-// Returns a subscription to a given doctype (all documents)
-export function subscribe(config, doctype, { docId, parse = doc => doc } = {}) {
-  if (!cozySocket) initCozySocket(config)
-  // Some document need to have specific parsing, for example, decoding
-  // base64 encoded properties
-  const parseCurried = listener => {
-    return doc => {
-      listener(parse(doc))
-    }
+  /**
+   * Remove the given handler from the list of handlers for given
+   * doctype/document and event.
+   *
+   * @param  {String}  type      Document doctype to subscribe to
+   * @param  {String}  id        Document id to subscribe to
+   * @param  {String}  eventName Event to subscribe to
+   * @param  {Function}  handler   Function to call when an event of the
+   * given type on the given doctype or document is received from stack.
+   * @return {Promise}           Promise that the message has been sent.
+   */
+  subscribe(eventName, type, handlerOrId, handlerOrUndefined) {
```

Comment:
Maybe the method signature should be clearer with id as last parameter, as @ptbrowne suggested.

I personally prefer a selector object as second parameter (who knows, maybe one day we will be happy to subscribe to realtime based on another criteria).

Reply:
J'ai jamais entendu @ptbrowne parler de mettre l'id en dernier. Ce code a été fait en parti avec lui.